src/coreComponents/mpiCommunications/CommunicationTools.cpp

```diff
@@ -748,21 +750,45 @@ void CommunicationTools::synchronizePackSendRecvSizes( const std::map< string, s
 }
 
 
-void CommunicationTools::synchronizePackSendRecv( const std::map< string, string_array > & fieldNames,
-                                                  MeshLevel & mesh,
-                                                  std::vector< NeighborCommunicator > & neighbors,
-                                                  MPI_iCommData & icomm,
-                                                  bool on_device )
+void CommunicationTools::asyncPack( const std::map< string, string_array > & fieldNames,
+                                    MeshLevel & mesh,
+                                    std::vector< NeighborCommunicator > & neighbors,
+                                    MPI_iCommData & icomm,
+                                    bool onDevice,
+                                    parallelDeviceEvents & events )
 {
   GEOSX_MARK_FUNCTION;
-
-  MPI_iCommData sizeComm;
+  if( onDevice )
+  {
+    localIndex eventCount = 0;
+    for( const auto & sa : fieldNames )
+    {
+      eventCount += sa.second.size( );
+    }
+    eventCount *= neighbors.size( );
+    events.reserve( eventCount );
```

Comment:
we are still packing each field independently?

Reply:
Yeah we're still launching packing kernels for each field for each type ('node','edge','face','elem') for each neighbor.

The sizing on this is pulled inside the neighbor communicator and more precise (this doesn't include the type count) in the follow-up I'm currently prepping.

Reply:
If this loop is only about pre-allocating enough space, I wouldn't bother with it. Calling `emplace_back` a handful (100s) of times on an empty vector or array1d should pale in comparison with either kernel launches or MPI calls, not to mention both.